File: studio-flow-invalid-definition/python/twilio_studio_flow_validity_audit.py
"""Report Twilio Studio Flows whose definition does not compile.

Read only. GET requests and nothing else: give this an API Key with read access
rather than the account auth token. The repair is printed, never performed,
because this script holds a credential to an account that can send messages and
spend money.
"""
import argparse
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("twilio_studio_flow_validity_audit")

STUDIO = "https://studio.twilio.com/v2"


def normalise(entries):
    """Reduce errors[] or warnings[] to deduplicated (path, message) pairs. Pure.

    Each entry names the widget in `path` and what broke in `message`. Entries
    that arrive as a bare string are kept with an empty path rather than
    dropped: a finding with no location is still a finding, and silently losing
    it is worse than printing it without one.

    The same fault is reported once per referencing transition, so a single
    deleted widget can produce four identical entries. Deduplicating here keeps
    the report a list of problems rather than a list of mentions.
    """
    out = []
    for e in entries or []:
        if isinstance(e, dict):
            path = str(e.get("path") or "").strip()
            message = str(e.get("message") or "").strip()
        else:
            path, message = "", str(e or "").strip()
        if not (path or message):
            continue
        pair = (path, message)
        if pair not in out:
            out.append(pair)
    return out


def verdict(flow):
    """Classify one Studio Flow by whether its definition compiles. Pure, so the
    five cases sit together instead of being spread through a request loop.

    `status` does not change the finding, only who is affected by it: a
    published Flow is failing executions now, a draft cannot be published until
    the widget is fixed. Returns (state, detail).
    """
    valid = flow.get("valid")
    status = str(flow.get("status") or "").lower()
    errors = normalise(flow.get("errors"))
    warnings = normalise(flow.get("warnings"))

    if valid is None:
        return ("unknown",
                "no valid field on this response: read the single flow at "
                "/v2/Flows/{FlowSid}, which is where errors[] and warnings[] "
                "are carried.")

    if valid is False:
        where = errors[0][0] if errors and errors[0][0] else "an unnamed widget"
        what = errors[0][1] if errors else "no message returned with the error"
        if not errors:
            detail = ("definition does not compile but errors[] came back empty. "
                      "Fetch the flow on its own; the list view is not where the "
                      "detail lives.")
        elif status == "published":
            detail = ("published and does not compile: executions stop at the "
                      "fault. %d error(s), first at %s: %s"
                      % (len(errors), where, what))
        else:
            detail = ("draft and does not compile, so it cannot be published at "
                      "all. %d error(s), first at %s: %s"
                      % (len(errors), where, what))
        return ("invalid-published" if status == "published" else "invalid-draft",
                detail)

    if warnings:
        return ("warnings",
                "compiles, with %d warning(s), first at %s: %s"
                % (len(warnings), warnings[0][0] or "an unnamed widget",
                   warnings[0][1] or "no message"))

    return ("valid", "definition compiles with no errors or warnings")


def get(session, url, **params):
    r = session.get(url, params=params, timeout=30)
    if r.status_code in (401, 403):
        raise SystemExit("%d from Twilio: check TWILIO_ACCOUNT_SID and that the "
                         "API key belongs to that account with read access"
                         % r.status_code)
    r.raise_for_status()
    return r.json()


def paged(session, url, key, limit, **first):
    """Page a studio.twilio.com list. meta.next_page_url is absolute."""
    params = dict(first)
    params.setdefault("PageSize", 50)
    out = []
    while url and len(out) < limit:
        page = get(session, url, **params)
        out.extend(page.get(key, []))
        url = (page.get("meta") or {}).get("next_page_url")
        params = {}
    return out[:limit]


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--max-flows", type=int, default=200,
                    help="stop after this many flows")
    ap.add_argument("--warnings", action="store_true",
                    help="also report flows that compile but carry warnings")
    args = ap.parse_args()

    account = os.environ.get("TWILIO_ACCOUNT_SID")
    key = os.environ.get("TWILIO_API_KEY")
    secret = os.environ.get("TWILIO_API_SECRET")
    if not (account and key and secret):
        log.error("set TWILIO_ACCOUNT_SID, TWILIO_API_KEY and TWILIO_API_SECRET "
                  "(an API Key with read access, not the auth token)")
        return 2

    session = requests.Session()
    session.auth = (key, secret)

    flows = paged(session, STUDIO + "/Flows", "flows", args.max_flows)
    if not flows:
        log.info("no Studio Flows on this account")
        return 0

    bad = 0
    for listed in flows:
        # The list gives the SIDs; errors[] and warnings[] come from the
        # single-flow fetch, and without them there is nothing to report but a
        # boolean saying something somewhere is wrong.
        flow = get(session, "%s/Flows/%s" % (STUDIO, listed.get("sid")))
        state, detail = verdict(flow)
        line = "%-18s %s  %s" % (state, flow.get("friendly_name") or flow.get("sid"),
                                 detail)

        if state == "valid":
            log.info(line)
            continue
        if state == "warnings" and not args.warnings:
            log.info("%-18s %s  %d warning(s); re-run with --warnings to see them",
                     state, flow.get("friendly_name") or flow.get("sid"),
                     len(normalise(flow.get("warnings"))))
            continue

        bad += 1
        log.warning(line)
        for path, message in normalise(flow.get("errors")):
            log.warning("  error at %s: %s", path or "(no path)", message)
        if state.startswith("invalid"):
            log.warning("  repair: fix the widget at that path in %s, check the "
                        "definition against %s/Flows/Validate, then republish.",
                        flow.get("sid"), STUDIO)

    log.info("%d flow(s), %d with a definition that does not compile",
             len(flows), bad)
    return 1 if bad else 0


if __name__ == "__main__":
    sys.exit(main())

File: studio-flow-invalid-definition/node/twilio-studio-flow-validity-audit.mjs
/**
 * Report Twilio Studio Flows whose definition does not compile.
 *
 * Read only. GET requests and nothing else: give this an API Key with read
 * access rather than the account auth token. The repair is printed, never
 * performed.
 */
const STUDIO = 'https://studio.twilio.com/v2';

/**
 * Reduce errors[] or warnings[] to deduplicated [path, message] pairs. Pure.
 *
 * Entries that arrive as a bare string keep an empty path rather than being
 * dropped. The same fault is reported once per referencing transition, so one
 * deleted widget can produce four identical entries; deduplicating here keeps
 * the report a list of problems rather than a list of mentions.
 */
export function normalise(entries) {
  const out = [];
  const seen = new Set();
  for (const e of entries ?? []) {
    let path = '';
    let message = '';
    if (e && typeof e === 'object') {
      path = String(e.path ?? '').trim();
      message = String(e.message ?? '').trim();
    } else {
      message = String(e ?? '').trim();
    }
    if (!path && !message) continue;
    const key = `${path}\u0000${message}`;
    if (seen.has(key)) continue;
    seen.add(key);
    out.push([path, message]);
  }
  return out;
}

/**
 * Classify one Studio Flow by whether its definition compiles. Pure.
 *
 * `status` does not change the finding, only who is affected by it: a published
 * Flow is failing executions now, a draft cannot be published until the widget
 * is fixed. Returns [state, detail].
 */
export function verdict(flow) {
  const valid = flow.valid;
  const status = String(flow.status ?? '').toLowerCase();
  const errors = normalise(flow.errors);
  const warnings = normalise(flow.warnings);

  if (valid === null || valid === undefined) {
    return ['unknown',
      'no valid field on this response: read the single flow at ' +
      '/v2/Flows/{FlowSid}, which is where errors[] and warnings[] are carried.'];
  }

  if (valid === false) {
    const where = errors.length && errors[0][0] ? errors[0][0] : 'an unnamed widget';
    const what = errors.length ? errors[0][1] : 'no message returned with the error';
    let detail;
    if (!errors.length) {
      detail = 'definition does not compile but errors[] came back empty. Fetch ' +
        'the flow on its own; the list view is not where the detail lives.';
    } else if (status === 'published') {
      detail = `published and does not compile: executions stop at the fault. ` +
        `${errors.length} error(s), first at ${where}: ${what}`;
    } else {
      detail = 'draft and does not compile, so it cannot be published at all. ' +
        `${errors.length} error(s), first at ${where}: ${what}`;
    }
    return [status === 'published' ? 'invalid-published' : 'invalid-draft', detail];
  }

  if (warnings.length) {
    return ['warnings',
      `compiles, with ${warnings.length} warning(s), first at ` +
      `${warnings[0][0] || 'an unnamed widget'}: ${warnings[0][1] || 'no message'}`];
  }

  return ['valid', 'definition compiles with no errors or warnings'];
}

function authHeader(key, secret) {
  return `Basic ${Buffer.from(`${key}:${secret}`).toString('base64')}`;
}

async function get(auth, url, params = {}) {
  const u = new URL(url);
  for (const [k, v] of Object.entries(params)) u.searchParams.set(k, v);
  const res = await fetch(u, { headers: { Authorization: auth } });
  if (res.status === 401 || res.status === 403) {
    throw new Error(`${res.status} from Twilio: check TWILIO_ACCOUNT_SID and ` +
                    'that the API key belongs to that account with read access');
  }
  if (!res.ok) throw new Error(`${res.status} from ${u.pathname}`);
  return res.json();
}

async function paged(auth, url, key, limit, first = {}) {
  let params = { PageSize: 50, ...first };
  const out = [];
  while (url && out.length < limit) {
    const page = await get(auth, url, params);
    out.push(...(page[key] ?? []));
    url = (page.meta ?? {}).next_page_url ?? null;
    params = {};
  }
  return out.slice(0, limit);
}

async function main() {
  const account = (process.env.TWILIO_ACCOUNT_SID || "dummy-twilio-account-sid");
  const key = (process.env.TWILIO_API_KEY || "dummy-twilio-api-key");
  const secret = (process.env.TWILIO_API_SECRET || "dummy-twilio-api-secret");
  if (!account || !key || !secret) {
    console.error('set TWILIO_ACCOUNT_SID, TWILIO_API_KEY and TWILIO_API_SECRET ' +
                  '(an API Key with read access, not the auth token)');
    process.exitCode = 2;
    return;
  }
  const auth = authHeader(key, secret);
  const showWarnings = process.argv.includes('--warnings');

  const flows = await paged(auth, `${STUDIO}/Flows`, 'flows', 200);
  if (!flows.length) {
    console.log('no Studio Flows on this account');
    return;
  }

  let bad = 0;
  for (const listed of flows) {
    const flow = await get(auth, `${STUDIO}/Flows/${listed.sid}`);
    const [state, detail] = verdict(flow);
    const name = flow.friendly_name || flow.sid;
    const line = `${state.padEnd(18)} ${name}  ${detail}`;

    if (state === 'valid') { console.log(line); continue; }
    if (state === 'warnings' && !showWarnings) {
      console.log(`${state.padEnd(18)} ${name}  ` +
                  `${normalise(flow.warnings).length} warning(s); re-run with ` +
                  '--warnings to see them');
      continue;
    }

    bad += 1;
    console.warn(line);
    for (const [path, message] of normalise(flow.errors)) {
      console.warn(`  error at ${path || '(no path)'}: ${message}`);
    }
    if (state.startsWith('invalid')) {
      console.warn(`  repair: fix the widget at that path in ${flow.sid}, check ` +
                   `the definition against ${STUDIO}/Flows/Validate, then republish.`);
    }
  }

  console.log(`${flows.length} flow(s), ${bad} with a definition that does not compile`);
  process.exitCode = bad ? 1 : 0;
}

// Only run when invoked directly. The test file imports this module, and without
// the guard main() would run there too, fail on the missing credentials and set a
// non-zero exit code that fails the whole test file even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
